third_party/maya/lib/pxrUsdMayaGL/proxyDrawOverride.cpp

```diff
@@ -194,11 +262,418 @@ UsdMayaProxyDrawOverride::supportedDrawAPIs() const
+					light.SetSpotCutoff(20.0);
+					light.SetSpotFalloff(0.0);
+					light.SetDiffuse(
+						GfVec4f(intensity * color[0], intensity * color[1], intensity * color[2], 1.0f));
+					light.SetAmbient(
+						GfVec4f(0.0f, 0.0f, 0.0f, 1.0f));
+					light.SetSpecular(
+						GfVec4f(1.0f, 1.0f, 1.0f, 1.0f));
+					// Spot light support shadow.
+					light.SetHasShadow(enableShadow);
+				}
+				else {
+					// Assumes a Maya Directional Light!
+					light.SetPosition(
+						GfVec4f(-direction[0], -direction[1], -direction[2], 0.0f));
+					light.SetSpotCutoff(180.0);
+					light.SetDiffuse(
+						GfVec4f(intensity * color[0], intensity * color[1], intensity * color[2], 1.0f));
+					light.SetAmbient(
+						GfVec4f(0.0f, 0.0f, 0.0f, 1.0f));
+					// Directional light support shadow.
+					light.SetHasShadow(enableShadow);
+				}
+			}
+			else if (hasPosition) {
+				// Assumes a Maya Point Light!
+				light.SetPosition(
+					GfVec4f(position[0], position[1], position[2], 1.0f));
+				// Maya's default value's for spot lights.
+				light.SetSpotCutoff(180.0);
+				light.SetDiffuse(
+					GfVec4f(intensity * color[0], intensity * color[1], intensity * color[2], 1.0f));
+				light.SetAmbient(
+					GfVec4f(0.0f, 0.0f, 0.0f, 1.0f));
+			}
+			else {
+				// Assumes a Maya Ambient Light!
+				light.SetPosition(
+					GfVec4f(0.0f, 0.0f, 0.0f, 1.0f));
+				light.SetSpotCutoff(180.0);
+				light.SetDiffuse(
+					GfVec4f(0.0f, 0.0f, 0.0f, 1.0f));
+				light.SetAmbient(
+					GfVec4f(intensity * color[0], intensity * color[1], intensity * color[2], 1.0f));
+			}
+
+			lights.push_back(light);
+		}
+	}
+
+	_GetBatchRenderer().SetLightings(lights);
+}
+
+void
+UsdMayaProxyDrawOverride::_RenderShapes(
+	const MHWRender::MDrawContext& context,
+	TfToken drawRepr,
+	bool enableLighting,
+	HdCullStyle cullStyle)
+{
+	// Draw only once for batch render in the same time stamp.
+	if (! _GetBatchRenderer().updateRenderTimeStamp(context.getFrameStamp()))
+		return;
+
+	// Get the world/view/projection matrix.
+	MStatus status;
+	MMatrix projectionMat = context.getMatrix(MHWRender::MDrawContext::kProjectionMtx, &status);
+	MMatrix viewMat = context.getMatrix(MHWRender::MDrawContext::kViewMtx, &status);
+	if (status != MStatus::kSuccess) return;
+
+	// Extract camera settings from maya view
+	int viewX, viewY, viewWidth, viewHeight;
+	context.getViewportDimensions(viewX, viewY, viewWidth, viewHeight);
+	GfVec4d viewport(viewX, viewY, viewWidth, viewHeight);
+
+	// Setup lighting and shadow before batch render.
+	bool enableShadow = false; // renderParams.shadow;
+	_SetupLighting(context, enableLighting, enableShadow);
+
+	// Batch render
+	//
+	_GetBatchRenderer().RenderBatches(
+		drawRepr,
+		cullStyle,
+		GfMatrix4d(viewMat.matrix),
+		GfMatrix4d(projectionMat.matrix),
+		viewport
+	);
+}
+
 void
 UsdMayaProxyDrawOverride::draw(const MHWRender::MDrawContext& context, const MUserData *data)
 {
-    _GetBatchRenderer().Draw( context, data );
+	// Draw callback
```

Comment:
Is it necessary that all of this moves out of the batch renderer and into the draw override here? This and a few other things that migrated into the draw override seem like they could be more generally useful if kept closer to the batch renderer. I think our hope is that the draw override delegates as much as possible to the batch renderer so that it's easier for shape and other plugin nodes that want to image using Hydra to leverage the batch renderer in their own draw overrides.

I know that there is currently a lot of stuff jammed into batchRenderer.h/.cpp, so I think migrating out the shape renderer and the task delegate to separate files makes a lot of sense, but I wonder whether it makes sense to keep more of the heavy lifting of the draw inside the batch renderer still.

Also, I know you'll hate to hear me say it, but we do still have a handful of use cases that necessitate using the legacy viewport from time to time, though we're working to get rid of those and get fully migrated to Viewport 2.0. In the meantime though, if this logic is moved out of the batch renderer and into the draw override, it would also need to be duplicated in the proxy shape UI that the legacy viewport uses.